### PR TITLE
 [DF] Fix #6853 -- issues with SaveGraph and Defines

### LINKED ISSUE #6853: [DF] SaveGraph can be confused by shared upstream Defines
## Describe the bug

Reported [on the forum](https://root-forum.cern.ch/t/crash-when-creating-rdataframe-graph/42343/3), with a reproducer.

[test3.C](https://root-forum.cern.ch/uploads/short-url/7ZVxQ0j0rgaCf65KgARmFplQGlz.C) + [testtree2.root](https://root-forum.cern.ch/uploads/short-url/bXK8KdWy4gp0xW8iT7zEIXAXJ3V.root) should produce 

image

but instead produce

image

when the following line is uncommented:

```
//latestDF = std::make_unique<RNode>(latestDF->Define("r42",[](){return 42;}));
```

Furthermore, `Define`s are displayed out of order, in general, by `SaveGraph`.

### Additional context

The problem is due to broken logic in the `GetGraph` function of `RAction` and `RFilter`: it loops over a `std::map` of Defines to figure out which have been added between that node and the previous node in the graph, but the loop over the `std::map` is out of order -- this causes both issues, the first because the out of order defines cause misattribution of parent nodes.

### PR DESCRIPTION
**Pull request by eguiraud**

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/RAction.hxx [hunk 1]
```diff
@@ -139,22 +139,28 @@ public:
...
-         if (std::find(prevColumns.begin(), prevColumns.end(), column.first) == prevColumns.end()) {
-            auto defineNode = RDFGraphDrawing::CreateDefineNode(column.first, column.second.get());
-            evaluatedNode->SetPrevNode(defineNode);
-            evaluatedNode = defineNode;
-         }
+      // Add the Defines that have been added between this node and the previous to the graph
+      auto upmostNode = thisNode;
+      const auto &defineNames = GetDefines().GetNames();
+      const auto &defineMap = GetDefines().GetColumns();
+      for (auto i = int(defineNames.size()) - 1; i >= 0; --i) {  // walk backwards through the names of defined columns
```
Comment (stwunsch) [inline tree/dataframe/inc/ROOT/RDF/RAction.hxx]: What's the complexity of this impl? Is it a N^2-like operation in every node of the graph?

**Review by stwunsch — APPROVED** [review]
LGTM! Just wonder about the complexity for searching through the lists (in every node?). Have we seen `SaveGraph` running smoothly on large RDF graphs?

File: tree/dataframe/inc/ROOT/RDF/RAction.hxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RAction.hxx]: no no, this checks just the defines that have been added between the previous non-Define node in the graph and this one -- due to the `break` 6 lines below.

And in general the chain of recursive `GetGraph` calls is cut short whenever we reach a node that was already visited.

Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RAction.hxx]: Alright, thanks! :+1:

**Pull request merged**